### ISSUE TITLE
Policy on open PRs

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#17683]
I notice that @czgdp1807 and @certik are reviewing open PRs with (I think) the aim of reducing the number of open PRs. I think this is a good thing because there are too many open PRs right now. However it seems to run against current practice so I think it would be good to discuss what the expected policy/workflow for open PRs is supposed to be.

We currently have around 600 open PRs with the oldest being #957 which is 7 years old. New PRs are opened all the time sometimes by new contributors and sometimes by experienced contributors. There are currently around 8000 closed PRs of which around 6000 have been merged. This suggests that roughly 3/4 PRs gets merged and around 10% of all PRs opened (since Github) were never merged or closed. Over the past year the new PRs were opened at an average of 3-4 per day so at that rate we could extrapolate that every 2-3 days a PR is opened that might never get merged or closed.

It seems to be currently accepted practice perhaps inspired by SymPy's "consensus" model that PRs are not generally closed except by the author. It often seems clear to me that a particular PR is unlikely to lead to anything useful but there is no agreed policy for simply closing it so I will leave it to accumulate with the others. This is inconsistent though with the idea that the list of open PRs represents a kind of TODO list that needs to be processed somehow - is that the model we want though?

If the plan is to reduce the number of open PRs then what should the criteria be for an author to keep a PR open? More importantly what should the criteria be for closing another author's PR?

Currently many contributors will leave a PR open even though it is unmergeable and they have no intention of making it mergeable. It seems like the expectation is that it should remain as an open PR until it has been superseded by a merged PR. My own #15449 is an example of this. I mean to find time for it but until then I imagine that simply leaving it there might serve as an inspiration for someone else to continue the work. Maybe I should just close the PR but make sure that the linked issue describes the work done and links to the closed PR.

**Comment by asmeurer** [comment]
We need some better way to indicate that a PR has useful work, but the author doesn't currently have time to work on it, and is OK if others continue that work, either by pushing up to the same PR (assuming they checked the box), or from a new PR. That also can happen for PRs where the author has disappeared, assuming the stuff that is there is useful. And more generally we should have a culture that it's OK to pick up someone else's work if they haven't been working on it. 

If the contents of the PR aren't useful in their current state, and the author has either disappeared or isn't able to work, we should be more willing to close. I think a big issue is that closing can give a bad impression on a PR author, that their contribution isn't desired or helpful. So we should craft some kind of response that we can use when closing, and put it in a template (it's too bad saved reply templates can't be shared across an organization). 

Also if relevant when closing a PR an issue should be opened if there is any work or discussion that shouldn't be lost.

**Comment by certik** [comment]
On Tue, Oct 1, 2019, at 4:21 PM, Oscar Benjamin wrote:
> I notice that @czgdp1807 <https://github.com/czgdp1807> and @certik 
> <https://github.com/certik> are reviewing open PRs with (I think) the 
> aim of reducing the number of open PRs. I think this is a good thing 
> because there are too many open PRs right now. However it seems to run 
> against current practice so I think it would be good to discuss what 
> the expected policy/workflow for open PRs is supposed to be.
> 
> We currently have around 600 open PRs with the oldest being #957 
> <https://github.com/sympy/sympy/pull/957> which is 7 years old. New PRs 
> are opened all the time sometimes by new contributors and sometimes by 
> experienced contributors. There are currently around 8000 closed PRs of 
> which around 6000 have been merged. This suggests that roughly 3/4 PRs 
> gets merged and around 10% of all PRs opened (since Github) were never 
> merged or closed. Over the past year the new PRs were opened at an 
> average of 3-4 per day so at that rate we could extrapolate that every 
> 2-3 days a PR is opened that might never get merged or closed.
> 
> It seems to be currently accepted practice perhaps inspired by SymPy's 
> "consensus" model that PRs are not generally closed except by the 
> author. It often seems clear to me that a particular PR is unlikely to 
> lead to anything useful but there is no agreed policy for simply 
> closing it so I will leave it to accumulate with the others. This is 
> inconsistent though with the idea that the list of open PRs represents 
> a kind of TODO list that needs to be processed somehow - is that the 
> model we want though?

I think we should change our old policy of leaving PRs open indefinitely to some other policy that allows us to manage our open PRs at a manageable level.


> 
> If the plan is to reduce the number of open PRs then what should the 
> criteria be for an author to keep a PR open? More importantly what 
> should the criteria be for closing another author's PR?

I am open to suggestions. The criteria can be as simple as:

* if the author is actively working on the PR (say it has commits from the last month), keep it open.
* if there was no commit, no comment for more than a month, and it's author's turn, then we can close it.  If the author gets back, he can open a new PR (because I think github does not allow the author to re-open his old PR).

Ondrej

**Comment by oscarbenjamin** [comment]
> I think github does not allow the author to re-open his old PR

I thought that an author could reopen a PR. At least they could use their branch to open a new PR otherwise.

It seems to me that the issue here is not so much whether the PR can be opened but that we don't want to discourage new contributors. If we close a PR then we need to give a clear, respectful explanation making clear that a contribution is appreciated even if not merged and that further contributions are welcome (including continuing the work from the closed PR).

**Comment by oscarbenjamin** [comment]
> I think we should change our old policy of leaving PRs open indefinitely to some other policy that allows us to manage our open PRs at a manageable level.

I assumed that was the case and I agree. I think though that the policy should be agreed and spelled out somewhere.

**Comment by oscarbenjamin** [comment]
> We need some better way to indicate that a PR has useful work, but the author doesn't currently have time to work on it, and is OK if others continue that work

@oscargus (I think) created the "easy to revive" label which is related to this:
https://github.com/sympy/sympy/pulls?q=is%3Aopen+is%3Apr+label%3A"Easy+to+Revive"

**Comment by certik** [comment]
Yes, let's spell it somewhere and then post a link there for every PR we close.

**Comment by asmeurer** [comment]
I might be wrong, but I don't think you can reopen a closed PR unless you have push access.

**Comment by sylee957** [comment]
There is a way to use label-exclusive searches like
`is:pr is:open -label:Documentation`
and in this way we can mark old PRs with some tag, and we would have to type in some exclusion flags for every searches though, but it can probably be a better way to manage some thread of stalled PRs in the community.

One problem about closing PR is that people would not likely notice that any existing work had been in the closed PRs. For the reason, I was sometimes annoyed by some authors closing his own PRs manually even if there were some valid move, conversely.

**Comment by czgdp1807** [comment]
Hi,
I suggest the following guidelines which I have formed taking into considerations the discussions held above:-
First of all let's divide the PRs into the following categories,

1. Active - These PRs are active and are receiving updates from the author and the reviewers daily or once in two-three days(in other words, frequency of updates is good).
2. In-active - These PRs are not active anymore and the last update(comment/commit) was given two to three months back.
3. Stalled - The update on these PRs happened more than two-three months back and we are unable to decide whether we should close it or continue it(with label `Could Close`) due to the changes in the master branch and merge conflicts.
4. Complete - The purpose of these PRs is complete(the associated issue is fixed).

How to handle the above categories?

1. Active PRs - We will keep reviewing and tracking them so that, they don't fall into in-active state. If author needs help we will try our best to help him out. Main goal will be to take the PR to mergable state. Recently opened PRs are good examples of this category.
2. In-active - We should ask the author that whether they are interested to continue their work on his/her PR. If he needs some help then we should provide that. If the author isn't interested then we can checkout their PRs and make a new one to fix the associated issue. In this way the author will get the credit and the patch will also get merged. https://github.com/sympy/sympy/pull/17673 is an example of continuing a previously inactive PR.
3. Stalled - In these kind of PRs we should ask author about his/her views on the current condition. If it was sympy's turn to provide help, then we should do it. Mostly, these PRs will be kept open till a conclusion is reached. Continuing these is a possibility if merge conflicts can be resolved(sometimes they get too complex).
4. Complete - If the author has completed his/her work and members support merging the PR then we should merge it. https://github.com/sympy/sympy/pull/12767 is an example of such a PR. Other case can be that the issue which suppose, PR A was trying to fix and PR B fixed it but forgot to automatically close PR A, then we will close the PR A with a nice message thanking the efforts of the author.

Please let me know your views on these guidelines. Note that there is only one case where a PR will be closed.

**Comment by oscarbenjamin** [comment]
@czgdp1807 I think that the way you are classifying things is fine but your workflow is missing the key points: how do we decide when to close a PR and how do we communicate that sensitively to the author?

**Comment by czgdp1807** [comment]
> how do we decide when to close a PR and how do we communicate that sensitively to the author?

I have mentioned in the `Complete` category about closing a PR. In all other cases, we should keep the PR open. About communicating to the author, we can tell them that the associated issue is fixed and we can share with them some more open issues which they can resolve through their new PRs. Let me know of your suggestions and ideas on this.

**Comment by oscarbenjamin** [comment]
Okay so what you are suggesting is that we don't close a PR until it has been superseded by another PR. That seems to be different from what (some) others in this thread are proposing.

**Comment by czgdp1807** [comment]
> if there was no commit, no comment for more than a month, and it's author's turn, then we can close it.

Well, we can do the above. I just suggested my point of view. What do you say? The ideal case is to keep the PRs active as much as possible, however sometimes it becomes difficult. What do you think, when should we close the PR?

**Comment by certik** [comment]
@czgdp1807 you will be doing most of this work, so I am fine with your classification, but I do suggest to modify your approach to actually close all PRs in the category 2, 3 and 4. Only keep the category 1 open. 

You can for example open an issue (and keep that issue open) for every PR that you close. And write in the issue what needs to be done to revive it.

But by only keeping the category 1 open, we will keep our PR tab manageable, which will tremendously help with new code getting in, as well as motivate authors of old code in old PRs to become active again.

**Comment by czgdp1807** [comment]
> You can for example open an issue (and keep that issue open) for every PR that you close. And write in the issue what needs to be done to revive it.

That's a great idea. I will wait for the author's response(for 24-72 hours), if none received, then I will either continue the PR(if it's in my domain) or close it and open an issue for it.

**Comment by certik** [comment]
@czgdp1807 perfect, great plan.

**Comment by certik** [comment]
P.S. You can think of it (and mention in the PRs) that you are "moving the PR from our PR tab into our Issues tab".

**Comment by oscarbenjamin** [comment]
I think it's important that we close the PRs that aren't being worked on and aren't mergeable.

The closure message should make clear that we will reopen it if the author wants to continue working on it and also should explain how the author can open a new PR from the same branch (assuming that's possible).

**Comment by asmeurer** [comment]
I'm not sure I agree that we should close PRs that aren't finished yet. That can send the wrong message that the work in the PR is not usable or that it isn't something we want done. If you look at my open PRs, https://github.com/sympy/sympy/pulls/asmeurer, all of them are things that I still think we should do. Most will require some additional work to finish. I'm happy if someone else picks up this work. I think this is less likely to happen if the PR is closed. It can be closed as soon as someone picks up the work and opens a new PR. And if the person picking up the work has push access, they can push directly to my branch (I always check the box to allow this), so a new PR is not even needed.  I know this somewhat "inflates" the PR count, but to me, these PRs are legitimately work in progress that should be finished, so it's correct that they are open. 

I completely agree that if the work in the PR isn't usable or desirable, or if it's clear that someone would have to restart from scratch, then we should close the PR.

**Comment by oscarbenjamin** [comment]
I'm not sure that this will send the wrong message if it is made clear both in the PR and in an open issue that the PR contains useful work but needs to be finished by someone. The important thing is the communication around this and what is said in the messages. I can also imagine that some contributors will not want to work on something *because* you have an open PR for it.

Rather than keep the PRs open would it not be better for you to close it with a message explaining that the work still needs doing and explaining why the approach you tried doesn't yet work? You can then open an issue flagging that not only have you stopped working on it but the fact that it needs finishing is an explicitly listed problem that needs to be solved.

**Comment by asmeurer** [comment]
I think a "can be picked up someone else" label can communicate this the best. 

Actually, if a PR has an approach that doesn't work at all, then that's something that should be closed and moved into an issue like you say. Because in that case, the work can't be used directly to fix the issue. The more common case is if the code works, but it needs to be cleaned up so that tests are fixed or certain review points addressed, or a merge conflict needs to be fixed. Often this isn't actually that much work to finish. 

By the way, there's also another possibility, which is a PR that the community is still deciding on the right way to do something, or if we want to do it at all. These should probably stay open and get the "needs decision" label until a consensus is reached. Or if the work in the PR proper isn't actually usable toward the goal regardless the discussion can be moved into an issue (like if it's a proof of concept PR *that is also stalled*).

**Comment by oscarbenjamin** [comment]
Okay so we classify inactive PRs like this:

1. The PR can be merged - merge it.
2. The PR can reasonably be continued to something mergeable - add the "please take over" label.
3. The PR takes a non-agreed approach - add the "needs decision" label.
4. The PR takes an approach that doesn't work or just isn't useful as a starting point - close it and open an issue explaining what should be done instead.
5. The PR itself isn't useful at all - just close it leaving any linked issue in place.

This way every inactive PR either gets merged, labelled, or closed.

Does that seem reasonable?

**Comment by asmeurer** [comment]
I would add for point 4 that if there is already an existing issue, it should just be added as a comment on that issue (but do make sure to add the comment; don't assume the automatic cross reference is enough). 

Otherwise it looks good. We also need good comments we can add to the PRs themselves, especially the ones that get closed, and we also need to agree on a time period after which a PR is considered stalled if no work has been done.

**Comment by certik** [comment]
We already have an existing document that outlines our current policy:

https://github.com/sympy/sympy/wiki/Closing-pull-requests

Can you please update it based on the above discussion? It also has a text to be put on PRs to be closed. We can improve the text if needed.

Let's document everything that we agreed upon, and let's start using the new policy.

**Comment by Sc0rpi0n101** [comment]
I think 30 days might be a good time period for an inactive PR. 
I would suggest if any pull request has no activity for more than 30 days, we add a comment asking the author if he can continue his work. 
If he does not reply within 24 hours, or indicates he cannot continue the work, we can tag the pull request accordingly.

**Comment by oscarbenjamin** [comment]
I think one month is fine for classifying a PR as inactive but 24hrs is too short to expect a response from the author. I certainly can't promise that I will always respond within 24hrs.

**Comment by Sc0rpi0n101** [comment]
We can give the author more time.
How much time would you suggest?

**Comment by certik** [comment]
3 days seems like a good time period. See our current policy I linked to above.

**Comment by Sc0rpi0n101** [comment]
3 days seems fine.
 
But, regarding the policy link, I could not find the link on the Wiki homepage or contributor documentation.
Since we plan to proactively monitor the pull requests, I suggest we also add the *Closing Pull Requests* and *Reviving Pull Requests* links where they are easily accessible to everyone.

**Comment by certik** [comment]
Yes, we should add it to sympy.org at a dedicated Development section.

**Comment by asmeurer** [comment]
I think we should keep it on the wiki. That's where we currently have all our developer policy resources. I agree it should be added to the main page (feel free to do so).

**Comment by czgdp1807** [comment]
I think we should go with this, https://github.com/sympy/sympy/issues/17683#issuecomment-537699712

**Comment by czgdp1807** [comment]
What about using the following PR flow,
![opencv-pr-flow](https://user-images.githubusercontent.com/36567889/68010269-2b251080-fcaa-11e9-8d06-22342e6a86e5.png)

References
[1] https://github.com/opencv/opencv/wiki/How_to_contribute#here-is-the-flow-chart-of-the-process

**Comment by czgdp1807** [comment]
Hi @certik @asmeurer 
I would like to add these two policies to SymPy wiki. Please let me know if it's okay to add them. I have been applying these two, to one of my projects and they seem work out pretty well but suggestions are welcome for making them better.

**Pull Request Policy**
```
The following should be done for any new Pull Request(PR) made to this repository,

1. The PR should be labelled with appropriate labels, can be done only by those who have write access.
2. Provide reviews - Make sure that the PR doesn't break any existing tests, coding style.
3. Wait period - Wait for the author to respond to the reviews. If they don't respond for continuously two weeks then attach a `Please take over` label to that PR.
4. If the PR is passing tests and there are no pending reviews then merge the PR, again only for those who have write/triage permissions.

#### When to create a new PR?

A contributor can only create a new PR for an open issue, not closed ones. However, if an issue is having more than one open PRs, then the best one is kept opened and the rest are closed in its favour. The following are the criteria for determining the best PR,

- Activity - If the author of the PR is active in maintaining the PR then it will be kept above other PRs.
- Test Status - If the tests are passing for a PR then that PR will be considered better.
- Coding Style - If the coding style is along the lines of the current master, then that PR will have higher chances of being kept opened.
- Recency - The more recent the PR is, the better it is.

#### When to close a PR?

If the reviewers feel that a PR doesn't has any scope of improvements and couldn't be merged even after providing reviews then they can either ask the maintainers to close the PR or they can close such PRs themselves if they have the access to do so.  

Some examples of the PRs that can be closed,

1. The code is in a language other than Python.
2. No reference to any issues or empty description in the PR.
3. Having things like driver code in the patch and looks like the contributor is spamming the pull request list.
4. Having irrelevant commits in the history from some other PR/branch.
5. Having features which are already there in the `master`.
6. Copied code without credits.
```

**Issue Policy**
```
We have the following policy for working on issues,

1. **No Assignment of Issues** - Issues are not assigned to contributors because we believe that concrete work matters more than a trivial assignment. Whenever a contributor makes a PR for an issue then that PR is linked with the issue and it reflects in the `Assignee` section of the issue tracker.

2. **How to claim issues?** - A contributor can comment on an issue to show his/her interest. If any maintainer responds positively then they can start the work according to the requirement of the issue. Note that, if the contributor doesn't respond for sometime with a comment or a PR then any other person can work on the same issue.

3. **Opening new issues** - If something isn't there in the master branch or there is a bug in the code base, anyone can open an issue for the same and work on it. Though the above points still apply on the author of the issue if he/she is interested to work on it. In addition, make sure that your issue doesn't duplicate any existing open/closed issue. 
```

Both of these policies are also available at the wiki of the project where I am applying these. Please let me know if there should be any modifications that are to be made for adding them to the wiki.

1. https://github.com/codezonediitj/pydatastructs/wiki/Pull-Request-Policy
2. https://github.com/codezonediitj/pydatastructs/wiki/Issue-Policy

**Comment by oscarbenjamin** [comment]
The wording is ambiguous e.g. "If any maintainer responds positively then they can start the work according to the requirement of the issue."

I don't think any permission should be needed to start working on something. I think it would be clearer just to say that anyone is free to work on any issue at any time. The problem with two people working on the same thing is that it is a duplication of effort so it is better to look for something that is not being worked on. The does not mean that it shouldn't happen in some situations though so no contributor should think that they own an issue.

**Comment by Arpan612** [comment]
@oscarbenjamin Sir, please have a look at the issue above. Even after repeatedly requesting the contributor to give me some time before jumping in, I have been ignored. I think my PR should be merged. I request your views on the same.

**Comment by czgdp1807** [comment]
The following are the points, 

1. More tests, for increasing the diff coverage, should be added in your patch. You can use pytest-cov or coverage.py to see the coverage locally. 
2. There are no policies which I am aware of in case of multiple PRs for the same issue. The discussions here are not finalised yet so I couldn't use the same for resolving the conflicts in your case. 
3. IMO, a PR can be merged by the approving comments from the community. I support merging #18960 as it appears better to me. If other members support merging your PR then no problem. At the end `master` will be having the required feature whichever PR is merged.
So, I would say let's wait for the comments from other members and then decide. You can ping the reviewers on your PR thread to receive feedback.

**Comment by Sc0rpi0n101** [comment]
> @oscarbenjamin Sir, please have a look at the issue above. Even after repeatedly requesting the contributor to give me some time before jumping in, I have been ignored. I think my PR should be merged. I request your views on the same.

I'm sorry, but your comment is not related to this issue. The contributors and maintainers work on their own time and availability. A Pull Request should be merged when the reviewers and the maintainers feel it is ready to be merged. But you cannot be spamming about it at unrelated places. 

Also, @czgdp1807 How about adding something about commits and commit messages before merging a pull request in the policy?
I've seen some odd and not too useful commit messages in the repo history and sometimes, some pull requests get merged with a large no of commits even for a very small change.  I think the maintainers should pay a little more attention to the commits before merging a Pull Request.

**Comment by czgdp1807** [comment]
> Also, @czgdp1807 How about adding something about commits and commit messages before merging a pull request in the policy?

IMO, that should be included. May be we can include it as one of the conditions(optional though) for merging a PR. For example, if there are too many commits then they should be squashed to one or just replaced with one commit by applying the latest diff after dropping all the commits. 

> some pull requests get merged with a large no of commits even for a very small change.

For resolving this may be we can activate, `Squash and Merge` feature of Github. It will automatically squash all the commits into one and then merge. The maintainers will be able to add their own commit message before merging. However, I am not sure if it changes the metadata and modifies the authors details after merging.

**Comment by czgdp1807** [comment]
Few updates on https://github.com/sympy/sympy/issues/17683#issuecomment-604294730
I have quoted the original one and updated version of that point is written below.
> **How to claim issues?** - A contributor can comment on an issue to show his/her interest. If any maintainer responds positively then they can start the work according to the requirement of the issue. Note that, if the contributor doesn't respond for sometime with a comment or a PR then any other person can work on the same issue.

**How to work on issues?** - A contributor can comment on an issue to show his interest and to obtain the latest status. However, any one can start working an any issue whenever they want to. It is recommended(though not forced) that contributors should not work on issues on which someone else is recently working.

> 4. If the PR is passing tests and there are no pending reviews then merge the PR, again only for those who have write/triage permissions.

4. If the PR is passing tests, there are no pending reviews and the commit history looks good then merge the PR, again only for those who have push access.

> #### When to close a PR?
> 
> If the reviewers feel that a PR doesn't has any scope of improvements and couldn't be merged even after providing reviews then they can either ask the maintainers to close the PR or they can close such PRs themselves if they have the access to do so.

#### When to add `Could Close` to a PR?

If the reviewers feel that a PR doesn't has any scope of improvements and couldn't be merged even after providing reviews then they can either ask the maintainers to add a `Could Close` label to the PR or they can do so by themselves if they have the access.
For closing PRs, the policy at https://github.com/sympy/sympy/wiki/Closing-pull-requests should be followed.

**Comment by czgdp1807** [comment]
Please let me know if any modifications are needed. I will try to make wiki pages within a week for these policies and will implement them after that. Thanks.

**Comment by sylee957** [comment]
> It is recommended(though not forced) that contributors should not work on issues on which someone else is recently working.

I don't think that it should be in the policy. Because it isn't a rule. But in a workflow thread.

**Comment by czgdp1807** [comment]
I have made the update.

**Comment by asmeurer** [comment]
A few notes:

> 1. The PR should be labelled with appropriate labels, can be done only by those who have write access.

This is not true. GitHub has a triage permission that we can use to give people permission to add labels without giving the write access. Already some people have triage permission. There is an idea to make the bot auto-give people Triage permissions (https://github.com/sympy/sympy-bot/issues/68).

Regarding issue assignment, there is also an idea to make the bot auto-assign people to issues when they open a PR. https://github.com/sympy/sympy-bot/issues/72. The idea here is that this would make it easier to see if a PR is already being worked on. However, we can debate whether this is something we really want to do. 

> 2. How to work on issues? - A contributor can comment on an issue to show his interest and to obtain the latest status. However, any one can start working an any issue whenever they want to. 

Use gender-neutral language here

> 3. **Opening new issues** - If something isn't there in the master branch or there is a bug in the code base, anyone can open an issue for the same and work on it. Though the above points still apply on the author of the issue if he/she is interested to work on it. In addition, make sure that your issue doesn't duplicate any existing open/closed issue. 

We should be consistent about third-person or second-person.

We should make it very clear (perhaps in bold) that people do not need to comment on an issue simply to ask if they can do it. Anyone can fix any issue without prior permission or assignment. The only consideration is if someone has already started work, in which case the other points apply.

**Comment by moorepants** [comment]
> It is recommended(though not forced) that contributors should not work on issues on which someone else is recently working.

We should encourage people to work together on issues. At a minimum we make authors aware of parallel work and at a maximum we help them merge their efforts.

**Comment by certik** [comment]
@czgdp1807 thanks for your efforts to document this and all your work you do in the PRs.

**Comment by moorepants** [comment]
Where can we see the current draft of this policy?

**Comment by czgdp1807** [comment]
The policy isn't ready yet. There is a policy on wiki about Closing PRs. I will add both Pull Request Policy and Issue Policy to the wiki by tomorrow. I am updating my comment where both of these are described according to the feedbacks received. 
In fact, both the policies will  always be in a constant state of improvement as anyone from the community will be able to edit them.

**Comment by moorepants** [comment]
That all sounds great, but this thread is hard to make sense of at this point. It would be nice to see a current draft of what we will agree to so we can comment on it. If you are willing to compile the current version I will be happy to review it.

**Comment by asmeurer** [comment]
I would put what we have on the wiki somewhere, but note at the top that it is a draft. That way we can more easily comment on it or just edit it.

**Comment by czgdp1807** [comment]
> GitHub has a triage permission that we can use to give people permission to add labels without giving the write access. Already some people have triage permission. There is an idea to make the bot auto-give people Triage permissions

That's a mistake in the policy. Thanks for pointing out. I will make the update.



> Regarding issue assignment, there is also an idea to make the bot auto-assign people to issues when they open a PR. [sympy/sympy-bot#72](https://github.com/sympy/sympy-bot/issues/72). The idea here is that this would make it easier to see if a PR is already being worked on

IMO, issues should not be assigned before making a PR and when a PR is linked with an issue, through, `Fixes`, `Closes`, etc., it reflects in the `Assignee` section of both PR and Issue tracker. Assigning issues before a PR is made for it will lead to stalled work restricting other contributors to explore.

> We should make it very clear (perhaps in bold) that people do not need to comment on an issue simply to ask if they can do it.

Sure, I will make the part highlighting this part bold.

**Comment by czgdp1807** [comment]
Both the policies are on wiki in draft state at the following links,

1. https://github.com/sympy/sympy/wiki/Pull-Request-Policy
2. https://github.com/sympy/sympy/wiki/Issue-Policy

Please let me know of any suggestions. I will try to implement them from next month when they will be mature enough and will be out of draft.

**Comment by moorepants** [comment]
> Provide reviews - Make sure that the PR doesn't break any existing tests, coding style.

Change to:

Provide reviews - Make sure that the PR doesn't break any existing tests, coding style is followed, deprecation policy is followed, and unit tests are added for new functionality and regressions testing.

**Comment by moorepants** [comment]
> If the PR is passing tests, there are no pending reviews and the commit history looks good then merge the PR, again only for those who have push access.

Not sure what "looks good" means. Maybe say "and the commit messages adequately describe the changes" instead.

**Comment by moorepants** [comment]
>  When to create a new PR?

This section does not encourage collaboration as I noted in this comment:

> We should encourage people to work together on issues. At a minimum we make authors aware of parallel work and at a maximum we help them merge their efforts.

I would rather see the wording of this change from determining what is the "best" pull request to language that directs merging efforts and doing so in a way that preserves commit contributions from the various pull requests.
